### PR TITLE
Feat: allow sweeping of other tokens & prevent sending ETH

### LINKED ISSUE #91: Allow sweeping of other tokens & prevent sending ETH
This issue describes 2 changes to the ClaimCODE.sol contract, as decided during our meeting on the 10th of May 2022.

### Ensure the contract cannot receive ETH

We want to ensure that the ClaimCODE contract is unable to receive ETH. Whether this was the default behaviour, or a specific revert in the fallback function has to be implemented is currently not known. Please first research/test whether a specific fallback function implementation is necessary.

Please note these breaking syntax changes of solidity 0.6: https://docs.soliditylang.org/en/latest/060-breaking-changes.html#semantic-and-syntactic-changes

### Allow to sweep any ERC20 tokens

To be able to sweep accidentally sent tokens to the contract, the `sweep` function should be extended to take an address as input.

If that address is the same as the `codeToken` contract's address, then the `block.timestamp <= claimPeriodEnds` requirement should still be checked. Other tokens could be swept before `claimPeriodEnds` as well.

### PR DESCRIPTION
**Pull request by zjhmale**
Closes #91 

1. add a `MockERC20` contract for testing emergency withdraw for other ERC20 tokens except $CODE.
2. verify `ClaimCODE` contract will never receive any Ethers, only contracts with `receive()` or `fallback()` function declared will embrace the ability to receive Ethers.

### CODE REVIEW

**Review by embiem — APPROVED** [review]
This looks good to me, no comments from my side!

**Comment by mariobyn** [comment]
Looks good to me as well

File: packages/hardhat/src/ClaimCODE.sol [hunk 1]
```diff
@@ -62,9 +62,9 @@ contract ClaimCODE is Ownable, Pausable {
         emit MerkleRootChanged(_merkleRoot);
     }
 
-    function sweep() external onlyOwner {
-        if (block.timestamp <= claimPeriodEnds) revert ClaimNotEnded();
-        codeToken.transfer(owner(), codeToken.balanceOf(address(this)));
+    function sweep(IERC20 token) external onlyOwner {
```
Comment (mariobyn) [inline packages/hardhat/src/ClaimCODE.sol:65]: i would say we should respect the notation of parameters, having _ in front of every param to differentiate the state variables from parameters.

**Comment by mariobyn** [comment]
One small comment, should we add a sweep version for ERC721 as well? making sure if people send their erc721 to the contract by mistake, can be recovered?

File: packages/hardhat/src/ClaimCODE.sol (hunk 1 above)
Comment (mariobyn) [reply packages/hardhat/src/ClaimCODE.sol:65]: also we miss an event emitted here, any state changes done to the contract, must emit an event.

**Comment by parseb** [comment]
> One small comment, should we add a sweep version for ERC721 as well? making sure if people send their erc721 to the contract by mistake, can be recovered?

We could. 
However, the codeclaim contract will be operational for a limited period of time, interaction will be overwhelmingly mediated by the fronted and its userbase is composed of 'web3 devs'. That being said I think this edge case is so unlikely that it does not merit the additional time and effort to implement it. (audit teams on standby) Do it if you think it's worth it.

File: packages/hardhat/src/ClaimCODE.sol (hunk 1 above)
Comment (zjhmale) [reply packages/hardhat/src/ClaimCODE.sol:65]: sure, will follow the rules!

**Comment by zjhmale** [comment]
agree with @parseb

**Comment by zjhmale** [comment]
I think sweep functions for the CODE contract is also worth equipping with :)

**Review by embiem — APPROVED** [review]

**Pull request merged**